File: benchmarks/kdtree-raycast/lang_go/kdtree.go
package main

import (
	"bufio"
	"encoding/binary"
	"math"
	"os"
	"unsafe"
)

const maxTraversalDepth = 64

type node [2]uint32

const (
	maxNodesCount        = 0x40000000
	leafNodeFlags uint32 = 3
)

func (n *node) initInteriorNode(axis int, aboveChild int32, split float32) {
	n[0] = uint32(axis) | uint32(aboveChild)<<2
	n[1] = *(*uint32)(unsafe.Pointer(&split))
}

func (n *node) initEmptyLeaf() {
	n[0] = leafNodeFlags // = 3
}

func (n *node) initLeafWithSingleTriangle(triangleIndex int32) {
	n[0] = leafNodeFlags | 1<<2 // = 7
	n[1] = uint32(triangleIndex)
}

func (n *node) initLeafWithMultipleTriangles(numTriangles int, triangleIndicesOffset int32) {
	// n[0] = 11, 15, 19, ... (for numTriangles = 2, 3, 4, ...)
	n[0] = leafNodeFlags | uint32(numTriangles)<<2
	n[1] = uint32(triangleIndicesOffset)
}

func (n node) isLeaf() bool {
	return n[0]&leafNodeFlags == leafNodeFlags
}

func (n node) isInteriorNode() bool {
	return !n.isLeaf()
}

func (n node) getLeafTrianglesCount() int {
	return int(n[0] >> 2)
}

func (n node) getLeafTriangleIndex() int {
	return int(n[1])
}

func (n node) getLeafTriangleIndicesOffset() int {
	return int(n[1])
}

func (n node) getInteriorNodeSplitAxis() int {
	return int(n[0] & leafNodeFlags)
}

func (n node) getInteriorNodeAboveChild() int32 {
	return int32(n[0] >> 2)
}

func (n node) getInteriorNodeSplit() float32 {
	return *(*float32)(unsafe.Pointer(&n[1]))
}

type KdTree struct {
	nodes           []node
	triangleIndices []int32
	mesh            *TriangleMesh
	meshBounds      BBox32
}

func NewKdTree(fileName string, mesh *TriangleMesh) *KdTree {
	file, err := os.Open(fileName)
	checkError(err)
	defer file.Close()

	reader := bufio.NewReader(file)

	var nodesCount int32
	err = binary.Read(reader, binary.LittleEndian, &nodesCount)
	checkError(err)

	nodes := make([]node, nodesCount)
	err = binary.Read(reader, binary.LittleEndian, &nodes)
	checkError(err)

	var triangleIndicesCount int32
	err = binary.Read(reader, binary.LittleEndian, &triangleIndicesCount)
	checkError(err)

	triangleIndices := make([]int32, triangleIndicesCount)
	err = binary.Read(reader, binary.LittleEndian, &triangleIndices)
	checkError(err)

	return &KdTree{
		nodes: nodes,
		triangleIndices: triangleIndices,
		mesh: mesh,
		meshBounds: mesh.GetBounds(),
	}
}

func (kdTree *KdTree) SaveToFile(fileName string) {
	file, err := os.Create(fileName)
	checkError(err)
	defer file.Close()

	writer := bufio.NewWriter(file)
	binary.Write(writer, binary.LittleEndian, int32(len(kdTree.nodes)))
	binary.Write(writer, binary.LittleEndian, kdTree.nodes)
	binary.Write(writer, binary.LittleEndian, int32(len(kdTree.triangleIndices)))
	binary.Write(writer, binary.LittleEndian, kdTree.triangleIndices)
	writer.Flush()
}

func (kdTree *KdTree) Intersect(ray *Ray) (intersection Intersection, hitFound bool) {
	meshBounds64 := BBox64FromBBox32(kdTree.meshBounds)
	tMin, tMax, intersectBounds := meshBounds64.Intersect(ray)
	if !intersectBounds {
		return
	}

	type traversalInfo struct {
		theNode *node
		tMin float64
		tMax float64
	}

	var traversalStack [maxTraversalDepth]traversalInfo
	traversalStackSize := 0

	theNode := &kdTree.nodes[0]

	var closestIntersection TriangleIntersection
	closestIntersection.t = math.Inf(+1)

	advanceNodePointer := func(n *node) (*node) {
		return (*node)(unsafe.Pointer(uintptr(unsafe.Pointer(n)) + 8))
	}

	for closestIntersection.t > tMin {
		if theNode.isInteriorNode() {
			axis := theNode.getInteriorNodeSplitAxis()
			distanceToSplitPlane :=float64(theNode.getInteriorNodeSplit()) - ray.GetOrigin()[axis]

			if distanceToSplitPlane == 0 {
				if ray.GetDirection()[axis] > 0 {
					theNode = &kdTree.nodes[theNode.getInteriorNodeAboveChild()]
				} else {
					theNode = advanceNodePointer(theNode)
				}
			} else {
				var firstChild, secondChild *node
				if distanceToSplitPlane > 0 {
					firstChild = advanceNodePointer(theNode)
					secondChild = &kdTree.nodes[theNode.getInteriorNodeAboveChild()]
				} else {
					firstChild = &kdTree.nodes[theNode.getInteriorNodeAboveChild()]
					secondChild = advanceNodePointer(theNode)
				}

				tSplit := distanceToSplitPlane * ray.GetInvDirection()[axis] // != 0
				if tSplit >= tMax || tSplit < 0 {
					theNode = firstChild
				} else if tSplit <= tMin {
					theNode = secondChild
				} else { // tMin < tSplit < tMax
					traversalStack[traversalStackSize] = traversalInfo{secondChild, tSplit, tMax}
					traversalStackSize++
					theNode = firstChild
					tMax = tSplit
				}
			}
		} else { // leaf node
			trianglesCount := theNode.getLeafTrianglesCount()
			if trianglesCount == 1 {
				triangleIndex := theNode.getLeafTriangleIndex()
				indices := kdTree.mesh.triangles[triangleIndex]
				triangle := Triangle{
					Vector64From32(kdTree.mesh.vertices[indices[0]]),
					Vector64From32(kdTree.mesh.vertices[indices[1]]),
					Vector64From32(kdTree.mesh.vertices[indices[2]]),
				}
				triangleIntersection, hitFound := IntersectTriangle(ray, &triangle)
				if hitFound && triangleIntersection.t < closestIntersection.t {
					closestIntersection = triangleIntersection
				}
			} else {
				for i:=0; i < trianglesCount; i++ {
					triangleIndex := kdTree.triangleIndices[theNode.getLeafTriangleIndicesOffset() + i]
					indices := kdTree.mesh.triangles[triangleIndex]
					triangle := Triangle{
						Vector64From32(kdTree.mesh.vertices[indices[0]]),
						Vector64From32(kdTree.mesh.vertices[indices[1]]),
						Vector64From32(kdTree.mesh.vertices[indices[2]]),
					}
					triangleIntersection, hitFound := IntersectTriangle(ray, &triangle)
					if hitFound && triangleIntersection.t < closestIntersection.t {
						closestIntersection = triangleIntersection
					}
				}
			}

			if traversalStackSize == 0 {
				break
			}

			traversalStackSize--
			theNode = traversalStack[traversalStackSize].theNode
			tMin = traversalStack[traversalStackSize].tMin
			tMax = traversalStack[traversalStackSize].tMax
		}
	}

	hitFound = closestIntersection.t < math.Inf(+1)
	if hitFound {
		intersection.RayT = closestIntersection.t
		intersection.RayTEpsilon = closestIntersection.rayEpsilon
	}
	return
}

File: benchmarks/kdtree-construction/lang_cpp/kdtree.cpp
#include "kdtree.h"
#include "triangle.h"

#include <cassert>
#include <fstream>

KdTree::KdTree(std::vector<Node>&& nodes,
               std::vector<int32_t>&& triangleIndices, const TriangleMesh& mesh,
               const BoundingBox_f& meshBounds)
    : _nodes(std::move(nodes)), _triangleIndices(std::move(triangleIndices)),
      _mesh(mesh), _meshBounds(meshBounds)
{
}

KdTree::KdTree(KdTree&& other)
    : _nodes(std::move(other._nodes)),
      _triangleIndices(std::move(other._triangleIndices)), _mesh(other._mesh),
      _meshBounds(other._meshBounds)
{
}

KdTree::KdTree(const std::string& kdtreeFileName, const TriangleMesh& mesh)
    : _mesh(mesh), _meshBounds(mesh.getBounds())
{
  std::ifstream file(kdtreeFileName, std::ios_base::in | std::ios_base::binary);
  if (!file)
    throw "KdTree: failed to load kdtree file";

  int32_t nodesCount;
  file.read(reinterpret_cast<char*>(&nodesCount), 4);
  if (!file)
    throw "KdTree: failed to read nodes count";

  auto& nodes = const_cast<std::vector<Node>&>(_nodes);
  nodes.resize(nodesCount);
  file.read(reinterpret_cast<char*>(nodes.data()), nodesCount * sizeof(Node));
  if (!file)
    throw "KdTree: failed to read kdtree nodes";

  int32_t triangleIndicesCount;
  file.read(reinterpret_cast<char*>(&triangleIndicesCount), 4);
  if (!file)
    throw "KdTree: failed to read triangle indices count";

  auto& triangleIndices = const_cast<std::vector<int32_t>&>(_triangleIndices);
  triangleIndices.resize(triangleIndicesCount);
  file.read(reinterpret_cast<char*>(triangleIndices.data()),
            triangleIndicesCount * 4);
  if (!file)
    throw "KdTree: failed to read triangle indices";
}

bool KdTree::intersect(const Ray& ray, Intersection& intersection) const
{
  auto boundsIntersection = _meshBounds.intersect(ray);
  if (!boundsIntersection.found)
    return false;

  struct TraversalInfo {
    const Node* node;
    double tMin;
    double tMax;
  };
  TraversalInfo traversalStack[maxTraversalDepth];
  int traversalStackSize = 0;

  double tMin = boundsIntersection.t0;
  double tMax = boundsIntersection.t1;

  auto node = &_nodes[0];

  RayTriangleIntersection closestIntersection;
  closestIntersection.t = std::numeric_limits<double>::infinity();

  while (closestIntersection.t > tMin) {
    if (node->isInteriorNode()) {
      int axis = node->getInteriorNodeSplitAxis();
      double distanceToSplitPlane = node->split - ray.getOrigin()[axis];

      if (distanceToSplitPlane == 0.0) {
        if (ray.getDirection()[axis] > 0.0)
          node = &_nodes[node->getInteriorNodeAboveChild()];
        else
          node += 1;
        // todo: check case when ray.direction[axis] == 0.0, probably in this
        // case we
        // we need to process both nodes to avoid rendering artifacts (although
        // rare cases)
      }
      else {
        const Node* firstChild;
        const Node* secondChild;

        if (distanceToSplitPlane > 0.0) {
          firstChild = node + 1;
          secondChild = &_nodes[node->getInteriorNodeAboveChild()];
        }
        else {
          firstChild = &_nodes[node->getInteriorNodeAboveChild()];
          secondChild = node + 1;
        }

        double tSplit =
            distanceToSplitPlane * ray.getInvDirection()[axis]; // != 0
        if (tSplit >= tMax || tSplit < 0.0)
          node = firstChild;
        else if (tSplit <= tMin)
          node = secondChild;
        else // tMin < tSplit < tMax
        {
          assert(traversalStackSize < maxTraversalDepth);
          traversalStack[traversalStackSize++] = {secondChild, tSplit, tMax};
          node = firstChild;
          tMax = tSplit;
        }
      }
    }
    else // leaf node
    {
      int trianglesCount = node->getLeafTrianglesCount();
      if (trianglesCount == 1) {
        auto indices = _mesh.triangles[node->index].vertexIndices;
        Triangle triangle;
        triangle.p[0] = Vector(_mesh.vertices[indices[0]]);
        triangle.p[1] = Vector(_mesh.vertices[indices[1]]);
        triangle.p[2] = Vector(_mesh.vertices[indices[2]]);

        RayTriangleIntersection triangleIntersection;
        bool intersectionFound =
            intersectTriangle(ray, triangle, triangleIntersection);
        if (intersectionFound &&
            triangleIntersection.t < closestIntersection.t) {
          closestIntersection = triangleIntersection;
        }
      }
      else {
        for (int32_t i = 0; i < trianglesCount; ++i) {
          auto indices =
              _mesh.triangles[_triangleIndices[node->index + i]].vertexIndices;
          Triangle triangle;
          triangle.p[0] = Vector(_mesh.vertices[indices[0]]);
          triangle.p[1] = Vector(_mesh.vertices[indices[1]]);
          triangle.p[2] = Vector(_mesh.vertices[indices[2]]);

          RayTriangleIntersection triangleIntersection;
          bool intersectionFound =
              intersectTriangle(ray, triangle, triangleIntersection);
          if (intersectionFound &&
              triangleIntersection.t < closestIntersection.t) {
            closestIntersection = triangleIntersection;
          }
        }
      }

      if (traversalStackSize == 0)
        break;

      --traversalStackSize;
      node = traversalStack[traversalStackSize].node;
      tMin = traversalStack[traversalStackSize].tMin;
      tMax = traversalStack[traversalStackSize].tMax;
    }
  }

  bool isHit = closestIntersection.t < std::numeric_limits<double>::infinity();
  if (isHit) {
    intersection.rayT = closestIntersection.t;
    intersection.rayTEpsilon = closestIntersection.rayEpsilon;
  }
  return isHit;
}

const TriangleMesh& KdTree::getMesh() const { return _mesh; }

const BoundingBox_f& KdTree::getMeshBounds() const { return _meshBounds; }

void KdTree::saveToFile(const std::string& fileName) const
{
  std::ofstream file(fileName, std::ios_base::out | std::ios_base::binary);
  if (!file)
    throw "Failed to write kdtree to file " + fileName;

  int32_t nodesCount = static_cast<int32_t>(_nodes.size());
  file.write(reinterpret_cast<const char*>(&nodesCount), 4);
  file.write(reinterpret_cast<const char*>(_nodes.data()),
             _nodes.size() * sizeof(Node));

  int32_t triangleIndicesCount = static_cast<int32_t>(_triangleIndices.size());
  file.write(reinterpret_cast<const char*>(&triangleIndicesCount), 4);
  file.write(reinterpret_cast<const char*>(_triangleIndices.data()),
             _triangleIndices.size() * 4);
}

uint64_t KdTree::getHash() const
{
  uint32_t hash1 = 0;
  for (const auto& node : _nodes) {
    hash1 = hash1 * 33 + node.header + node.index;
  }

  uint32_t hash2 = 0;
  for (int32_t index : _triangleIndices) {
    hash2 = hash2 * 33 + uint32_t(index);
  }

  uint64_t hash = hash1 + (static_cast<uint64_t>(hash2) << 32);
  return hash;
}
